From fd6166cd228c4c1b65173d9b112eeee07ddccaf7 Mon Sep 17 00:00:00 2001
From: chandrasekhar <cchaduvu@amd.com>
Date: Tue, 17 Mar 2026 22:47:38 +0530
Subject: [PATCH] init_sensor: add sensor_address to support runtime sensor
 driver switching

Enable runtime switching of sensor drivers by allowing the active sensor I2C address
to be updated dynamically during initialization.The serializer I2C alias mapping is
reprogrammed when a valid link is present so that subsequent I2C transactions are
redirected to the selected sensor address.

Signed-off-by: chandrasekhar <cchaduvu@amd.com>
---
 .../source/Multi_sensor_framework/ms_fw.c     | 47 ++++++++++++++++++-
 .../src/usr_sensor/source/sensor_drv/imx623.c |  2 +-
 .../src/usr_sensor/source/sensor_drv/imx728.c | 21 ++-------
 .../usr_sensor/source/sensor_drv/ox03f10.c    |  2 +-
 .../usr_sensor/source/sensor_drv/ox05b1s.c    |  2 +-
 .../usr_sensor/source/sensor_drv/ox08b40.c    |  2 +-
 .../usr_sensor/source/sensor_drv/sensor_drv.c |  2 +-
 7 files changed, 54 insertions(+), 24 deletions(-)

diff --git a/sensor_drivers/src/usr_sensor/source/Multi_sensor_framework/ms_fw.c b/sensor_drivers/src/usr_sensor/source/Multi_sensor_framework/ms_fw.c
index b1495de..feabba2 100755
--- a/sensor_drivers/src/usr_sensor/source/Multi_sensor_framework/ms_fw.c
+++ b/sensor_drivers/src/usr_sensor/source/Multi_sensor_framework/ms_fw.c
@@ -25,6 +25,7 @@
  ******************************************************************************/
 
 #include "FreeRTOS.h"
+#include "hal_i2c.h"
 #include "task.h"
 #include "queue.h"
 #include "timers.h"
@@ -259,7 +260,7 @@ void init_iic_access(u8 i2cBusId, int in_pipe)
  * @return  None
  *
  *****************************************************************************/
-void init_sensor(int in_pipe, int des_index)
+void init_sensor(int in_pipe, int des_index, int sensor_address)
 {
 	xil_printf("enter init_sensor \r\n");
 	/*
@@ -340,6 +341,50 @@ void init_sensor(int in_pipe, int des_index)
 				(des_arr[des_index].link_b.sensor_alias_addr);
 		}
 	}
+
+	if (des_arr[des_index].link_type != NO_LINK) {
+		xil_printf("Sensor address passed: %x\n",sensor_address);
+		RESULT Status = RET_SUCCESS;
+		u16 reg_addr;
+		u8 ser_addr = g_fmc_single.serializer_array[in_pipe]->alias_addr >> 1;
+		u8 wr_data[2] = {0};
+		u8 read_data[2] = {0};
+		
+		reg_addr = I2C_2;
+		Status = HalReadI2CReg(IIC_INSTANCE_ZERO, ser_addr, reg_addr, 0x2, read_data, 1);
+		
+		if (Status != XST_SUCCESS) DCT_ASSERT(0);
+//		xil_printf("before I2c_2 :%x\r\n", read_data[0]);
+		
+		reg_addr = I2c_3;
+		Status = HalReadI2CReg(IIC_INSTANCE_ZERO, ser_addr, reg_addr, 0x2, read_data, 1);
+		
+		if (Status != XST_SUCCESS) DCT_ASSERT(0);
+//		xil_printf("before I2c_3 :%x\r\n", read_data[0]);
+
+		reg_addr = I2C_2;
+		wr_data[0] = g_fmc_single.sensor_array[in_pipe]->sensor_alias_addr;
+		Status = HalWriteI2CReg(IIC_INSTANCE_ZERO, ser_addr, reg_addr, 0x2, wr_data[0], 1); 
+		if (Status != XST_SUCCESS) DCT_ASSERT(0);
+		
+		reg_addr = I2C_3;
+		wr_data[0] = sensor_address;
+		Status = HalWriteI2CReg(IIC_INSTANCE_ZERO, ser_addr, reg_addr, 0x2, wr_data[0], 1); 
+		if (Status != XST_SUCCESS) DCT_ASSERT(0);
+		
+		
+		reg_addr = I2C_2;
+		Status = HalReadI2CReg(IIC_INSTANCE_ZERO, ser_addr, reg_addr, 0x2, read_data, 1);
+		
+		if (Status != XST_SUCCESS) DCT_ASSERT(0);
+//		xil_printf("After I2c_2 :%x\r\n", read_data[0]);
+		
+		reg_addr = I2c_3;
+		Status = HalReadI2CReg(IIC_INSTANCE_ZERO, ser_addr, reg_addr, 0x2, read_data, 1);
+		
+		if (Status != XST_SUCCESS) DCT_ASSERT(0);
+//		xil_printf("After I2c_3 :%x\r\n", read_data[0]);
+	}
 }
 
 /*******************************************************************************
diff --git a/sensor_drivers/src/usr_sensor/source/sensor_drv/imx623.c b/sensor_drivers/src/usr_sensor/source/sensor_drv/imx623.c
index 8883646..d01e959 100755
--- a/sensor_drivers/src/usr_sensor/source/sensor_drv/imx623.c
+++ b/sensor_drivers/src/usr_sensor/source/sensor_drv/imx623.c
@@ -394,7 +394,7 @@ static RESULT IMX623_IsiCreateIss(IsiSensorInstanceConfig_t *pConfig, IsiSensorH
 
 	init_iic_access(pIMX623Ctx->i2cId, pipeId);
 	init_des(desId);
-	init_sensor(pipeId, desId);
+	init_sensor(pipeId, desId, SENSOR_IMX623_ADDRESS);
 
 	TRACE(IMX623_INFO, "%s (exit)\n", __func__);
 	return result;
diff --git a/sensor_drivers/src/usr_sensor/source/sensor_drv/imx728.c b/sensor_drivers/src/usr_sensor/source/sensor_drv/imx728.c
index 5538fc3..98f4f94 100755
--- a/sensor_drivers/src/usr_sensor/source/sensor_drv/imx728.c
+++ b/sensor_drivers/src/usr_sensor/source/sensor_drv/imx728.c
@@ -34,8 +34,8 @@
 #include "isi/isi_priv.h"
 #include "sensor_drv/imx728_priv.h"
 
-#define WITH_VTS		(1)
-//#define WITH_HTS		(0)
+// #define WITH_VTS		(1)
+#define WITH_HTS		(1)
 
 extern int g_Sensor_frame_count;
 
@@ -398,7 +398,7 @@ static RESULT IMX728_IsiCreateIss(IsiSensorInstanceConfig_t *pConfig, IsiSensorH
 	TRACE(IMX728_INFO, "desId %d pipeId:%d\n", desId, pIMX728Ctx->sensorDevId);
 
 	init_des(desId);
-	init_sensor(pIMX728Ctx->sensorDevId, desId);
+	init_sensor(pIMX728Ctx->sensorDevId, desId, SENSOR_IMX728_ADDRESS);
 	init_iic_access(pIMX728Ctx->i2cId, pIMX728Ctx->sensorDevId);
 
 	return result;
@@ -1491,26 +1491,11 @@ static RESULT IMX728_IsiSetStreamingIss(IsiSensorHandle_t handle,  bool_t mode)
 			return RET_FAILURE;
 		}
 	} else {
-		result = IMX728_IsiWriteRegIss(handle,  IMX728_REG_STREAM, 0xA3);
-		if (result != RET_SUCCESS) {
-			TRACE(IMX728_ERROR, "%s: Failed to start sensor stream!\n", __func__);
-			return RET_FAILURE;
-		}
 		result = IMX728_IsiWriteRegIss(handle,  IMX728_REG_STREAM, 0xFF);
 		if (result != RET_SUCCESS) {
 			TRACE(IMX728_ERROR, "%s: Failed to Stop sensor stream!\n", __func__);
 			return RET_FAILURE;
 		}
-		result = IMX728_IsiWriteRegIss(handle,  IMX728_REG_SLEEP, 0x52);
-		if (result != RET_SUCCESS) {
-			TRACE(IMX728_ERROR, "%s: Failed to Stop sensor stream!\n", __func__);
-			return RET_FAILURE;
-		}
-		result = IMX728_IsiWriteRegIss(handle,  IMX728_REG_SLEEP, 0xAE);
-		if (result != RET_SUCCESS) {
-			TRACE(IMX728_ERROR, "%s: Failed to Stop sensor stream!\n", __func__);
-			return RET_FAILURE;
-		}
 	}
 
 	pIMX728Ctx->streaming = mode;
diff --git a/sensor_drivers/src/usr_sensor/source/sensor_drv/ox03f10.c b/sensor_drivers/src/usr_sensor/source/sensor_drv/ox03f10.c
index 5564a6f..9f13a90 100755
--- a/sensor_drivers/src/usr_sensor/source/sensor_drv/ox03f10.c
+++ b/sensor_drivers/src/usr_sensor/source/sensor_drv/ox03f10.c
@@ -378,7 +378,7 @@ static RESULT Ox03f10_IsiCreateIss(IsiSensorInstanceConfig_t *pConfig, IsiSensor
 
 	init_iic_access(pOx03f10Ctx->i2cId, pipeId);
 	init_des(desId);
-	init_sensor(pipeId, desId);
+	init_sensor(pipeId, desId, SENSOR_OX3F10_ADDRESS);
 
 	TRACE(Ox03f10_INFO, "%s (exit)\n", __func__);
 	return result;
diff --git a/sensor_drivers/src/usr_sensor/source/sensor_drv/ox05b1s.c b/sensor_drivers/src/usr_sensor/source/sensor_drv/ox05b1s.c
index 8c4fff6..9f14b14 100755
--- a/sensor_drivers/src/usr_sensor/source/sensor_drv/ox05b1s.c
+++ b/sensor_drivers/src/usr_sensor/source/sensor_drv/ox05b1s.c
@@ -421,7 +421,7 @@ static RESULT Ox05b1s_IsiCreateIss(IsiSensorInstanceConfig_t *pConfig, IsiSensor
 
 	if (mcmABmode_initCount <= 0) {
 		init_des(desId);
-		init_sensor(pipeId, desId);
+		init_sensor(pipeId, desId, SENSOR_OX5B_ADDRESS);
 	}
 	mcmABmode_initCount++;
 
diff --git a/sensor_drivers/src/usr_sensor/source/sensor_drv/ox08b40.c b/sensor_drivers/src/usr_sensor/source/sensor_drv/ox08b40.c
index 0cce73d..ab2dd89 100755
--- a/sensor_drivers/src/usr_sensor/source/sensor_drv/ox08b40.c
+++ b/sensor_drivers/src/usr_sensor/source/sensor_drv/ox08b40.c
@@ -400,7 +400,7 @@ static RESULT Ox08b40_IsiCreateIss(IsiSensorInstanceConfig_t *pConfig, IsiSensor
 
 	init_iic_access(pOx08b40Ctx->i2cId, pipeId);
 	init_des(desId);
-	init_sensor(pipeId, desId);
+	init_sensor(pipeId, desId, SENSOR_OX8B40_ADDRESS);
 
 	TRACE(Ox08b40_INFO, "%s (exit)\n", __func__);
 	return result;
diff --git a/sensor_drivers/src/usr_sensor/source/sensor_drv/sensor_drv.c b/sensor_drivers/src/usr_sensor/source/sensor_drv/sensor_drv.c
index 65c0a28..f07e74e 100755
--- a/sensor_drivers/src/usr_sensor/source/sensor_drv/sensor_drv.c
+++ b/sensor_drivers/src/usr_sensor/source/sensor_drv/sensor_drv.c
@@ -225,7 +225,7 @@ RESULT SensorDrvGetPortInfo(sensorPortInfo_t *pPortInfo, uint32_t sensorDevId)
 
 	if (Status != XST_SUCCESS)
 		return Status;
-	init_sensor(pipeId, desId);
+	init_sensor(pipeId, desId, 0);
 
 	for (uint8_t index = 0; index < (ARRAY_SIZE(sensorCfgList)); index++) {
 
-- 
2.43.0

